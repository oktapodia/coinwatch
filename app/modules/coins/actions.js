import settings from 'electron-settings';
import { find, findIndex } from 'lodash';
import { CALL_HANDLER } from '../../middleware/handler';
import { getCoinListApi, getCoinPriceApi, getExchangeListApi, getSymbolListApi } from '../../connectors/cryptocompare/api';
import { generateSlug } from './helpers';

export const GET_COINS_ATTEMPT = 'GET_COINS_ATTEMPT';
export const GET_COINS_SUCCESS = 'GET_COINS_SUCCESS';
export const GET_COINS_FAILURE = 'GET_COINS_FAILURE';

export function getCoins() {
  return {
    [CALL_HANDLER]: {
      types: [GET_COINS_ATTEMPT, GET_COINS_SUCCESS, GET_COINS_FAILURE],
      handler: getCoinListApi,
    },
  };
}

export const GET_COIN_PRICE_ATTEMPT = 'GET_COIN_PRICE_ATTEMPT';
export const GET_COIN_PRICE_SUCCESS = 'GET_COIN_PRICE_SUCCESS';
export const GET_COIN_PRICE_FAILURE = 'GET_COIN_PRICE_FAILURE';

export function getCoinPrice(coin) {
  return {
    [CALL_HANDLER]: {
      types: [GET_COIN_PRICE_ATTEMPT, GET_COIN_PRICE_SUCCESS, GET_COIN_PRICE_FAILURE],
      handler: getCoinPriceApi,
      data: coin,
    },
  };
}

export const GET_EXCHANGE_LIST_ATTEMPT = 'GET_EXCHANGE_LIST_ATTEMPT';
export const GET_EXCHANGE_LIST_SUCCESS = 'GET_EXCHANGE_LIST_SUCCESS';
export const GET_EXCHANGE_LIST_FAILURE = 'GET_EXCHANGE_LIST_FAILURE';

export function getExchangeList() {
  return {
    [CALL_HANDLER]: {
      types: [GET_EXCHANGE_LIST_ATTEMPT, GET_EXCHANGE_LIST_SUCCESS, GET_EXCHANGE_LIST_FAILURE],
      handler: getExchangeListApi,
    },
  };
}

export const GET_SYMBOL_LIST_ATTEMPT = 'GET_SYMBOL_LIST_ATTEMPT';
export const GET_SYMBOL_LIST_SUCCESS = 'GET_SYMBOL_LIST_SUCCESS';
export const GET_SYMBOL_LIST_FAILURE = 'GET_SYMBOL_LIST_FAILURE';

export function getSymbolList() {
  return {
    [CALL_HANDLER]: {
      types: [GET_SYMBOL_LIST_ATTEMPT, GET_SYMBOL_LIST_SUCCESS, GET_SYMBOL_LIST_FAILURE],
      handler: getSymbolListApi,
    },
  };
}

export const SETTINGS_SAVE_COIN_ATTEMPT = 'SETTINGS_SAVE_COIN_ATTEMPT';
export const SETTINGS_SAVE_COIN_SUCCESS = 'SETTINGS_SAVE_COIN_SUCCESS';
export const SETTINGS_SAVE_COIN_FAILED = 'SETTINGS_SAVE_COIN_FAILED';

export function saveCoin(coin) {
  return (dispatch) => {
    dispatch({ type: SETTINGS_SAVE_COIN_ATTEMPT });

    if (!settings.has('coins')) {
      settings.set('coins', []);
    }

    // eslint-disable-next-line
    coin.slug = generateSlug(coin.coin.Symbol, coin.to, coin.exchange);

    const coins = settings.get('coins');

    if (find(coins, (c) => c.slug === coin.slug)) {
      return dispatch({ type: SETTINGS_SAVE_COIN_FAILED });
    }

    // eslint-disable-next-line
    coin.visibility = false;

    coins.push(coin);
    settings.set('coins', coins);

    return dispatch({ type: SETTINGS_SAVE_COIN_SUCCESS });
  };
}

export const SETTINGS_SAVE_ALERT_ATTEMPT = 'SETTINGS_SAVE_ALERT_ATTEMPT';
export const SETTINGS_SAVE_ALERT_SUCCESS = 'SETTINGS_SAVE_ALERT_SUCCESS';
export const SETTINGS_SAVE_ALERT_FAILED = 'SETTINGS_SAVE_ALERT_FAILED';

export function saveAlert(alert) {
  return (dispatch) => {
    dispatch({ type: SETTINGS_SAVE_ALERT_ATTEMPT });

    const alerts = settings.get('alerts');

    console.log(alert);

    alerts[alert.slug] = alert;

    const index = findIndex(alerts, (c) => c.slug === alert.slug);
    if (index !== -1) {
      alerts.splice(index, 1);
    }

    alerts.push(alert);

    console.log('updated', alerts);

    settings.set('alerts', alerts);

    return dispatch({ type: SETTINGS_SAVE_ALERT_SUCCESS });
  };
}

export const SETTINGS_REMOVE_COIN_ATTEMPT = 'SETTINGS_REMOVE_COIN_ATTEMPT';
export const SETTINGS_REMOVE_COIN_SUCCESS = 'SETTINGS_REMOVE_COIN_SUCCESS';
export const SETTINGS_REMOVE_COIN_FAILED = 'SETTINGS_REMOVE_COIN_FAILED';

export function removeCoin(slug) {
  return (dispatch) => {
    dispatch({ type: SETTINGS_REMOVE_COIN_ATTEMPT });

    const coins = settings.get('coins');
    const index = findIndex(coins, (c) => c.slug === slug);
    if (index === -1) {
      return dispatch({ type: SETTINGS_REMOVE_COIN_FAILED });
    }

    coins.splice(index, 1);

    settings.set('coins', coins);

    return dispatch({ type: SETTINGS_REMOVE_COIN_SUCCESS });
  };
}

export const SETTINGS_TOGGLE_VISIBILITY_SUCCESS = 'SETTINGS_TOGGLE_VISIBILITY_SUCCESS';

export function toggleVisibility(slug) {
  return (dispatch) => dispatch({ type: SETTINGS_TOGGLE_VISIBILITY_SUCCESS, data: { slug } });
}

export const FORCE_REFRESH_TOGGLE = 'FORCE_REFRESH_TOGGLE';

export function toggleForceRefresh() {
  return (dispatch) => dispatch({ type: FORCE_REFRESH_TOGGLE });
}
