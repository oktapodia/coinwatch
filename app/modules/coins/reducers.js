import settings from 'electron-settings';
import { cloneDeep, find, map, findIndex } from 'lodash';
import { ipcRenderer } from 'electron';
import {
  FORCE_REFRESH_TOGGLE,
  GET_COIN_PRICE_SUCCESS,
  GET_COINS_SUCCESS,
  GET_EXCHANGE_LIST_SUCCESS,
  GET_SYMBOL_LIST_SUCCESS,
  SETTINGS_REMOVE_COIN_SUCCESS,
  SETTINGS_SAVE_COIN_SUCCESS,
  SETTINGS_TOGGLE_VISIBILITY_SUCCESS,
  SETTINGS_SAVE_ALERT_SUCCESS,
} from './actions';
import { TRAY_UPDATE } from '../../main/Tray';
import { NOTIFICATION } from '../../main/NotificationCenter';

const TREND_LOWER = 'lower';
const TREND_HIGHER = 'higher';

const initialState = {
  data: {},
  exchanges: [],
  symbols: [],
  coins: settings.get('coins') || [],
  previousCoins: [],
  forceRefresh: false,
  alertsSettings: settings.get('alerts') || [],
  alerts: {},
};

export default function coinsReducer(state = initialState, action) {
  switch (action.type) {
    case SETTINGS_TOGGLE_VISIBILITY_SUCCESS: {
      const coins = cloneDeep(state.coins);
      const coin = find(coins, (c) => c.slug === action.data.slug);

      coin.visibility = !coin.visibility;

      settings.set('coins', coins);

      return { ...state, coins };
    }
    case SETTINGS_SAVE_ALERT_SUCCESS:
      return { ...state, alertsSettings: settings.get('alerts') };
    case SETTINGS_SAVE_COIN_SUCCESS:
      return { ...state, coins: settings.get('coins') };
    case SETTINGS_REMOVE_COIN_SUCCESS:
      return { ...state, coins: settings.get('coins') };
    case GET_COINS_SUCCESS:
      return { ...state, data: action.response };
    case GET_EXCHANGE_LIST_SUCCESS:
      return { ...state, exchanges: action.response };
    case GET_SYMBOL_LIST_SUCCESS:
      return { ...state, symbols: action.response };
    case FORCE_REFRESH_TOGGLE:
      return { ...state, forceRefresh: !state.forceRefresh };
    case GET_COIN_PRICE_SUCCESS: {
      const coins = cloneDeep(state.coins);
      const previousCoins = cloneDeep(state.coins);
      const coin = find(coins, (c) => c.slug === action.data.slug);
      const previousCoin = find(state.previousCoins, (c) => c.slug === action.data.slug);

      if (!coin) {
        return { ...state };
      }

      coin.price = map(action.response).join('');

      if (previousCoin && coin.price < previousCoin.price) {
        coin.trend = TREND_LOWER;
      } else if (previousCoin && coin.price > previousCoin.price) {
        coin.trend = TREND_HIGHER;
      } else {
        coin.trend = previousCoin && previousCoin.trend;
      }

      ipcRenderer.send(TRAY_UPDATE, coins);

      const response = { ...state, coins, previousCoins, alerts: state.alerts };

      // ------------------------------------------------------------------------------------------------------Alert
      const alerts = cloneDeep(state.alerts);

      const alertSetting = find(state.alertsSettings, (a) => a.slug === action.data.slug);
      if (!alertSetting) {
        return response;
      }

      if (!alerts[action.data.slug]) {
        alerts[action.data.slug] = null;
      }

      alerts[action.data.slug] = coin.price;

      response.alerts = alerts;

      let isAlerted = false;
      if (alertSetting.trend === 'down' && coin.price < alertSetting.price) {
        ipcRenderer.send(NOTIFICATION, { type: 'alert', title: coin.slug, body: `Price is lower than ${coin.price}` });
        isAlerted = true;
      } else if (alertSetting.trend === 'up' && coin.price > alertSetting.price) {
        ipcRenderer.send(NOTIFICATION, { type: 'alert', title: coin.slug, body: `Price is greater than ${coin.price}` });
        isAlerted = true;
      }

      if (isAlerted) {
        const index = findIndex(state.alertsSettings, (s) => s.slug === coin.slug);
        state.alertsSettings.splice(index, 1);

        settings.set('alerts', state.alertsSettings);
      }

      return response;
    }
    default:
      return state;
  }
}
