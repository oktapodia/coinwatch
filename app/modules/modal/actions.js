export const OPEN_MODAL = 'OPEN_MODAL';

export function openModal(component, extras = {}) {
  return (dispatch) => {
    dispatch({ type: OPEN_MODAL, component, extras });
  };
}

export const CLOSE_MODAL = 'CLOSE_MODAL';

export function closeModal() {
  return (dispatch) => {
    dispatch({ type: CLOSE_MODAL });
  };
}
