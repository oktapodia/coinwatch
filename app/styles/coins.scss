.coins {
  thead > tr {
    background-color: #333;

    a {
      color: #fff;

      &:hover {
        color: #ccc;
      }
    }
  }

  .coin {
    background-color: $table-tr-background-color;
    color: #000;
  }

  .coin:nth-of-type(odd) {
    background-color: $table-tr-background-color-odd;
  }

  .coin {
    //display: flex;

    & > div {
      flex-grow: 1;
    }

    .name {
      //display: flex;
      //flex-direction: row;
      //text-align: left;
      //flex-grow: 2;

      img {
        width: 30px;
        height: 30px;
      }

      span {
        margin-left: 5px;
        margin-top: 5px;
      }
    }

    .price {
      //display: flex;
      //flex-direction: row;
      margin: 5px 50px;

      &.trend- {
        &higher {
          color: green;
        }
        &lower {
          color: red;
        }
        &equal {
          color: blue;
        }
      }
    }

    .actions {
      margin-top: 5px;
      color: red;
      text-align: right;

      & > a {
        margin: 0 5px;
      }

      .remove {
        color: grey;

        &:hover {
          color: red;
        }
      }

      .alert {
        color: grey;

        &:hover {
          color: green;
        }
      }
    }
  }
}

form {
  input[type="submit"],
  input[type="submit"]:active {
    background: $navbar-color;
    color: #fff;
    margin: 10px auto;
    width: 100%;

    &:hover {
      background: lighten($navbar-color, 15);
      color: #fff;
    }
  }
}
