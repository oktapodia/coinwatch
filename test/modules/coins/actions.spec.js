import * as actions from '../../../src/renderer/modules/coins/actions';

describe('actions', () => {
  it('should get the coins list', () => {
    expect(actions.getCoins()).toMatchSnapshot();
  });

  it('should get the coin price with a coin argument only', () => {
    const data = {
      coin: {
        Symbol: 'ETH',
      },
    };
    expect(actions.getCoinPrice(data)).toMatchSnapshot();
  });

  it('should get the coin price with a coin and symbol arguments', () => {
    const data = {
      coin: {
        Symbol: 'ETH',
      },
      symbol: 'USD',
    };
    expect(actions.getCoinPrice(data)).toMatchSnapshot();
  });

  it('should get the coin price with a coin and symbol and exchange arguments', () => {
    const data = {
      coin: {
        Symbol: 'ETH',
      },
      symbol: 'USD',
      exchange: 'Bitfinex',
    };
    expect(actions.getCoinPrice(data)).toMatchSnapshot();
  });
});
