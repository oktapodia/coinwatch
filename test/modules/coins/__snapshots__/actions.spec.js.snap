// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions should get the coin price with a coin and symbol and exchange arguments 1`] = `
Object {
  Symbol(Call Handler): Object {
    "data": Object {
      "coin": Object {
        "Symbol": "ETH",
      },
      "exchange": "Bitfinex",
      "symbol": "USD",
    },
    "handler": [Function],
    "types": Array [
      "GET_COIN_PRICE_ATTEMPT",
      "GET_COIN_PRICE_SUCCESS",
      "GET_COIN_PRICE_FAILURE",
    ],
  },
}
`;

exports[`actions should get the coin price with a coin and symbol arguments 1`] = `
Object {
  Symbol(Call Handler): Object {
    "data": Object {
      "coin": Object {
        "Symbol": "ETH",
      },
      "symbol": "USD",
    },
    "handler": [Function],
    "types": Array [
      "GET_COIN_PRICE_ATTEMPT",
      "GET_COIN_PRICE_SUCCESS",
      "GET_COIN_PRICE_FAILURE",
    ],
  },
}
`;

exports[`actions should get the coin price with a coin argument only 1`] = `
Object {
  Symbol(Call Handler): Object {
    "data": Object {
      "coin": Object {
        "Symbol": "ETH",
      },
    },
    "handler": [Function],
    "types": Array [
      "GET_COIN_PRICE_ATTEMPT",
      "GET_COIN_PRICE_SUCCESS",
      "GET_COIN_PRICE_FAILURE",
    ],
  },
}
`;

exports[`actions should get the coins list 1`] = `
Object {
  Symbol(Call Handler): Object {
    "handler": [Function],
    "types": Array [
      "GET_COINS_ATTEMPT",
      "GET_COINS_SUCCESS",
      "GET_COINS_FAILURE",
    ],
  },
}
`;

exports[`actions should get the exchanges list 1`] = `
Object {
  Symbol(Call Handler): Object {
    "handler": [Function],
    "types": Array [
      "GET_EXCHANGE_LIST_ATTEMPT",
      "GET_EXCHANGE_LIST_SUCCESS",
      "GET_EXCHANGE_LIST_FAILURE",
    ],
  },
}
`;
